packages/nextjs/src/server/keyless-node.ts

```diff
@@ -0,0 +1,177 @@
+  try {
+    const CONFIG_PATH = getKeylessConfigurationPath();
+    let fileAsString;
+    try {
+      fileAsString = readFileSync(CONFIG_PATH, { encoding: 'utf-8' }) || '{}';
+    } catch {
+      fileAsString = '{}';
+    }
+    return JSON.parse(fileAsString) as AccountlessApplication;
+  } catch {
+    return undefined;
+  }
+}
+
+const createMessage = (keys: AccountlessApplication) => {
+  return `\n\x1b[35m\n[Clerk]:\x1b[0m You are running on keyless mode.\nYou can \x1b[35mclaim your keys\x1b[0m by visiting ${keys.claimUrl}\n`;
+};
+
+async function createOrReadKeyless(): Promise<AccountlessApplication | undefined> {
+  if (!nodeRuntime.fs) {
+    // This should never happen.
+    throw "Clerk: fsModule.fs is missing. This is an internal error. Please contact Clerk's support.";
+  }
+  const { existsSync, writeFileSync, mkdirSync, rmSync } = nodeRuntime.fs;
+
+  /**
+   * If another request is already in the process of acquiring keys return early.
+   * Using both an in-memory and file system lock seems to be the most effective solution.
+   */
+  if (isCreatingFile || existsSync(CLERK_LOCK)) {
+    return undefined;
+  }
+
+  isCreatingFile = true;
+
+  writeFileSync(
+    CLERK_LOCK,
+    // In the rare case, the file persists give the developer enough context.
+    'This file can be deleted. Please delete this file and refresh your application',
+    {
+      encoding: 'utf8',
+      mode: '0777',
+      flag: 'w',
+    },
+  );
+
+  const CONFIG_PATH = getKeylessConfigurationPath();
+  const README_PATH = getKeylessReadMePath();
+
+  mkdirSync(nodeRuntime.path ? nodeRuntime.path.dirname(CONFIG_PATH) : '', { recursive: true });
+  mkdirSync(nodeRuntime.path ? nodeRuntime.path.dirname(README_PATH) : '', { recursive: true });
+  updateGitignore();
+
+  /**
+   * When the configuration file exists, always read the keys from the file
+   */
+  const envVarsMap = safeParseClerkFile();
+  if (envVarsMap?.publishableKey && envVarsMap?.secretKey) {
+    isCreatingFile = false;
+    rmSync(CLERK_LOCK, { force: true, recursive: true });
+
+    /**
+     * Notify developers.
+     */
+    logger.logOnce(createMessage(envVarsMap));
+
+    return envVarsMap;
+  }
+
+  /**
+   * At this step, it is safe to create new keys and store them.
+   */
+  const client = createClerkClientWithOptions({});
+  const accountlessApplication = await client.__experimental_accountlessApplications.createAccountlessApplication();
+
+  /**
+   * Notify developers.
+   */
+  logger.logOnce(createMessage(accountlessApplication));
+
+  writeFileSync(CONFIG_PATH, JSON.stringify(accountlessApplication), {
+    encoding: 'utf8',
+    mode: '0777',
+    flag: 'w',
+  });
+
+  // TODO-KEYLESS: Add link to official documentation.
+  const README_NOTIFICATION = `
+## DO NOT COMMIT
+This file is auto-generated from \`@clerk/nextjs\` because you are running on Keyless mode. Avoid commiting the \`.clerk/\` directory as it includes the secret key of the unclaimed instance.
```

Comment:
```suggestion
This directory is auto-generated from \`@clerk/nextjs\` because you are running in Keyless mode. Avoid committing the \`.clerk/\` directory as it includes the secret key of your unclaimed instance.
```